File: deployment-services/shared/resilience/http_client.py
import requests
import logging
from typing import Any, Optional, Dict
from .circuit_breaker import CircuitBreaker

logger = logging.getLogger("resilience")

class ResilientHttpClient:
    def __init__(
        self,
        name: str,
        base_url: Optional[str] = None,
        timeout: float = 5.0,
        circuit_breaker_options: Optional[Dict[str, Any]] = None
    ):
        self.name = name
        self.base_url = base_url
        self.timeout = timeout
        
        cb_opts = circuit_breaker_options or {}
        self.breaker = CircuitBreaker(name, **cb_opts)
        self.session = requests.Session()

    def get(self, url: str, **kwargs) -> requests.Response:
        return self._execute("GET", url, **kwargs)

    def post(self, url: str, **kwargs) -> requests.Response:
        return self._execute("POST", url, **kwargs)

    def put(self, url: str, **kwargs) -> requests.Response:
        return self._execute("PUT", url, **kwargs)

    def patch(self, url: str, **kwargs) -> requests.Response:
        return self._execute("PATCH", url, **kwargs)

    def delete(self, url: str, **kwargs) -> requests.Response:
        return self._execute("DELETE", url, **kwargs)

    def _execute(self, method: str, url: str, **kwargs) -> requests.Response:
        full_url = f"{self.base_url.rstrip('/')}/{url.lstrip('/')}" if self.base_url else url
        
        if "timeout" not in kwargs:
            kwargs["timeout"] = self.timeout

        def make_request():
            res = self.session.request(method, full_url, **kwargs)
            res.raise_for_status()
            return res

        start_time = requests.compat.time.time()
        try:
            response = self.breaker.execute(make_request)
            duration = requests.compat.time.time() - start_time
            logger.info(f"HTTP {method} {full_url} succeeded - status: {response.status_code}, duration: {duration:.3f}s")
            return response
        except requests.exceptions.HTTPError as e:
            if e.response.status_code >= 500:
                logger.error(f"HTTP {method} {full_url} failed with server error: {e}")
                raise
            return e.response
        except Exception as e:
            logger.error(f"HTTP {method} {full_url} failed: {e}")
            raise

File: identity-services/packages/common/src/resilience/http-client.ts
import axios, { type AxiosInstance, type AxiosRequestConfig, type AxiosResponse } from "axios";
import { CircuitBreaker, type CircuitBreakerOptions, type CircuitMetrics } from "./circuit-breaker";
import { CreateLogger } from "../utils/logger";

const rootLogger = CreateLogger({ name: "resilience" });

export interface ResilientHttpClientOptions {
    /** Base URL for all requests */
    baseURL?: string;
    /** Default request timeout in ms (default: 5000) */
    requestTimeout?: number;
    /** Name for this client – used in logs and circuit breaker label */
    name: string;
    /** Circuit breaker options */
    circuitBreaker?: CircuitBreakerOptions;
    /** Additional axios defaults */
    axiosDefaults?: AxiosRequestConfig;
}
export class ResilientHttpClient {
    private readonly client: AxiosInstance;
    private readonly breaker: CircuitBreaker;
    private readonly requestTimeout: number;
    private readonly log: ReturnType<typeof rootLogger.child>;

    constructor(options: ResilientHttpClientOptions) {
        this.requestTimeout = options.requestTimeout ?? 5_000;
        this.log = rootLogger.child({ httpClient: options.name });

        this.client = axios.create({
            baseURL: options.baseURL,
            timeout: this.requestTimeout,
            ...options.axiosDefaults,
        });

        this.breaker = new CircuitBreaker(options.name, options.circuitBreaker);
    }

    async get<T = unknown>(
        url: string,
        config?: AxiosRequestConfig
    ): Promise<AxiosResponse<T>> {
        return this.execute<T>(() =>
            this.client.get<T>(url, this.withTimeout(config))
        );
    }

    async post<T = unknown>(
        url: string,
        data?: unknown,
        config?: AxiosRequestConfig
    ): Promise<AxiosResponse<T>> {
        return this.execute<T>(() =>
            this.client.post<T>(url, data, this.withTimeout(config))
        );
    }

    async put<T = unknown>(
        url: string,
        data?: unknown,
        config?: AxiosRequestConfig
    ): Promise<AxiosResponse<T>> {
        return this.execute<T>(() =>
            this.client.put<T>(url, data, this.withTimeout(config))
        );
    }

    async patch<T = unknown>(
        url: string,
        data?: unknown,
        config?: AxiosRequestConfig
    ): Promise<AxiosResponse<T>> {
        return this.execute<T>(() =>
            this.client.patch<T>(url, data, this.withTimeout(config))
        );
    }

    async delete<T = unknown>(
        url: string,
        config?: AxiosRequestConfig
    ): Promise<AxiosResponse<T>> {
        return this.execute<T>(() =>
            this.client.delete<T>(url, this.withTimeout(config))
        );
    }

    getCircuitMetrics(): CircuitMetrics {
        return this.breaker.getMetrics();
    }

    private async execute<T>(fn: () => Promise<AxiosResponse<T>>): Promise<AxiosResponse<T>> {
        const start = Date.now();
        return this.breaker.execute(async () => {
            try {
                const res = await fn();
                this.log.info(
                    { status: res.status, durationMs: Date.now() - start },
                    "HTTP request succeeded"
                );
                return res;
            } catch (err) {
                this.log.error(
                    { err, durationMs: Date.now() - start },
                    "HTTP request failed"
                );
                throw err;
            }
        });
    }

    private withTimeout(config?: AxiosRequestConfig): AxiosRequestConfig {
        return { timeout: this.requestTimeout, ...config };
    }
}
